src/lib.rs

```diff
@@ -1388,6 +1424,140 @@ where
+where
+    TE: ErrorResponse,
+    TIR: TokenInspectionResponse<TT>,
+    TT: TokenType,
+{
+    token: &'a AccessToken,
+    token_type_hint: Option<Cow<'a, str>>,
+
+    auth_type: &'a AuthType,
+    client_id: &'a ClientId,
+    client_secret: Option<&'a ClientSecret>,
+    extra_params: Vec<(Cow<'a, str>, Cow<'a, str>)>,
+    introspect_url: Option<&'a IntrospectUrl>,
+
+    _phantom: PhantomData<(TE, TIR, TT)>,
+}
+
+impl<'a, TE, TIR, TT> IntrospectRequest<'a, TE, TIR, TT>
+where
+    TE: ErrorResponse + 'static,
+    TIR: TokenInspectionResponse<TT>,
+    TT: TokenType,
+{
+    ///
+    /// Sets the optional token_type_hint parameter
+    ///
+    /// See: https://tools.ietf.org/html/rfc7662#section-2.1
+    ///
+    /// OPTIONAL.  A hint about the type of the token submitted for
+    ///       introspection.  The protected resource MAY pass this parameter to
+    ///       help the authorization server optimize the token lookup.  If the
+    ///       server is unable to locate the token using the given hint, it MUST
+    ///      extend its search across all of its supported token types.  An
+    ///      authorization server MAY ignore this parameter, particularly if it
+    ///      is able to detect the token type automatically.  Values for this
+    ///      field are defined in the "OAuth Token Type Hints" registry defined
+    ///      in OAuth Token Revocation [RFC7009](https://tools.ietf.org/html/rfc7009).
+    ///
+    pub fn set_token_type_hint<V>(mut self, value: V) -> Self
+    where
+        V: Into<Cow<'a, str>>,
+    {
+        self.token_type_hint = Some(value.into());
+
+        self
+    }
+
+    ///
+    /// Appends an extra param to the token introspect.
+    ///
+    /// This method allows extensions to be used without direct support from
+    /// this crate. If `name` conflicts with a parameter managed by this crate, the
+    /// behavior is undefined. In particular, do not set parameters defined by
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749) or
+    /// [RFC 7662](https://tools.ietf.org/html/rfc7662).
+    ///
+    /// # Security Warning
+    ///
+    /// Callers should follow the security recommendations for any OAuth2 extensions used with
+    /// this function, which are beyond the scope of
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749).
+    ///
+    pub fn add_extra_param<N, V>(mut self, name: N, value: V) -> Self
+    where
+        N: Into<Cow<'a, str>>,
+        V: Into<Cow<'a, str>>,
+    {
+        self.extra_params.push((name.into(), value.into()));
+        self
+    }
+
+    fn prepare_request<RE>(self) -> Result<HttpRequest, RequestTokenError<RE, TE>>
+    where
+        RE: Error + 'static,
+    {
+        let mut params: Vec<(&str, &str)> = vec![("token", self.token.secret())];
+        if let Some(ref token_type_hint) = self.token_type_hint {
+            params.push(("token_type_hint", token_type_hint));
+        }
+
+        Ok(endpoint_request(
+            self.auth_type,
```

Comment:
the spec leaves the details around client auth somewhat open-ended:
>  To prevent token scanning attacks, the endpoint MUST also require
   some form of authorization to access this endpoint, such as client
   authentication as described in OAuth 2.0 [RFC6749] or a separate
   OAuth 2.0 access token such as the bearer token described in OAuth
   2.0 Bearer Token Usage [RFC6750].  The methods of managing and
   validating these authentication credentials are out of scope of this
   specification.

it would probably be worth adding a `set_auth_method` (or similar) function to choose between:
1. client id+secret auth (either in the `Authorization` header or request body, as determined by `self.auth-type`)
2. specifying a bearer token (see the first example in [Sec. 2.1](https://tools.ietf.org/html/rfc7662#section-2.1))
3. no automatic auth added by this crate (clients can either add their own custom headers by using a custom `http_client`, or we could add a corresponding `add_request_header` function similar to `add_extra_param`)

however, using client id+secret auth seems like a sensible default, and we should be able to add `set_auth_method` in the future without introducing a breaking change, so it's not a requirement for this PR.

Reply:
I started implementing but then I saw this library already used the `AuthType`, so I re-used that.
AuthType could indeed be extended with a bearer token option in a later phase

Reply:
I had been thinking that `AuthType` should only support the auth methods used by the token endpoint, but checking [RFC 6749](https://tools.ietf.org/html/rfc6749#section-2.3) again, I see:
> The authorization server MAY accept any form of client authentication meeting its security requirements.

given this, I'm fine with just using `AuthType` and extending it in the future. I pushed 3ccf4252da1ab81c80074296347a762456174132 to make it `#[non_exhaustive]`.